File: CS315-Programming Languages/HW1/Codes/22101782_akin_yasemin.lua
-- 22101782_akin_yasemin.lua

-- Boolean expressions in Lua

-- 1. Boolean operators
local a = true
local b = false
print("a = ", a)
print("b = ", b)

print("1. Boolean operators in Lua:")
print("AND 'and'(a and b): ", a and b) -- false
print("OR 'or' (a or b): ", a or b) -- true
print("NOT 'not' (not a): ", not a) -- false

-- 2. Data types for operands of boolean operators, Boolean values (Truthy and Falsy values)
print("\n2. Data types for operands of boolean operators, Boolean values (Truthy and Falsy values) in Lua: in Lua, only 'false' and 'nil' are 'falsy'. Everything else is 'truthy'.")
print("Boolean value of 1: ", 1 and true or false) -- true ('truthy' value)
print("Boolean value of nil: ", nil and true or false) -- false ('falsy' value)
print("Boolean value of 0: ", 0 and true or false) -- true (0 is 'truthy')
print("Boolean value of an empty string: ", "" and true or false) -- true (empty string is 'truthy')

-- 3. Operator precedence rules
print("\n3. Operator precedence rules in Lua: not > and > or")
local c = false
print("a or b and not c: ", a or b and not c)  -- Equivalent to (a or (b and (not c)))

local d = true
print("a or b == c and d ~= a", a or b == c and d ~= a) -- Equivalent to (a or ((b == c) and (d ~= a)))

print("not a and b or c == d and d ~= a: ", not a and b or c == d and d ~= a) -- Equivalent to (((not a) and b) or ((c == d) and (d ~= a)))
-- ((false and b) or ((c == d) and (d != a)))
-- ((false and b) or (false and false))
-- (false or false)
-- false

-- 4. Operator associativity rules
print("\n4. Operator associativity rules in Lua: left-associative")
print("a and b and c and d: ", a and b and c and d) -- Evaluates as (((a and b) and c) and d)

local x = 0
local y = 1
print("x > y or y > x or y == x: ", x > y or y > x or y == x) -- Evaluates as ((x > y or y > x) or y == x)

-- 5. Operand evaluation order
print("\n5. Operand evaluation order in Lua: left-to-right")
local function first_method()
  print("First method called")
  return true
end

local function second_method()
  print("Second method called")
  return false
end

if first_method() and second_method() then end  -- Output: "First method called" then "Second method called"
if second_method() or first_method() then end -- Outputs: "Second method called" then "First method called"

-- 6. Short-circuit evaluation
print("\n6. Short-circuit evaluation:")
-- Lua uses short-circuit evaluation for 'and' and 'or'.

-- For 'and', if the first operand is false, the second operand is not evaluated.
local result = b and second_method()   -- "Second method called" is not printed
print("Result of b and second_method(): ", result)

-- For 'or', if the first operand is true, the second operand is not evaluated.
result = a or second_method()    -- "Second method called" is not printed
print("Result of a or second_method(): ", result)

File: CS315-Programming Languages/HW1/Codes/22101782_akin_yasemin.rb
# 22101782_akin_yasemin.rb

# Boolean expressions in Ruby

# 1. Boolean operators
a = true
b = false
puts "a = #{a}"
puts "b = #{b}"

puts "1. Boolean operators in Ruby:"
puts "AND '&&'(a && b): #{a && b}" # false
puts "OR '||' (a || b): #{a || b}" # true
puts "NOT '!' (!a): #{!a}" # false

# 2. Data types for operands of boolean operators, Boolean values
puts "\n2. Data types for operands of boolean operators, Boolean values in Ruby: in Ruby, almost every value is 'truthy', except for 'false' and 'nil' which are 'falsy'."
# 'true' and 'false' are the only boolean values. Everything else is truthy, except 'nil'.
puts !!1 # true ('truthy' value)
puts !!nil # false ('falsy' value)
puts !!0 # true (0 is 'truthy')
puts !"" # false
non_bool = "I am not a boolean"
if non_bool
    puts "In Ruby, non-boolean types can be used in boolean contexts. This will print."
end
number = 5
if number
    puts "Numbers other than 0 are also truthy in Ruby. This will print."
end

# 3. Operator precedence rules
puts "\n3. Operator precedence rules in Ruby: ! > && > ||"
c = false
puts "a || b && !c: #{a || b && !c}"  # Equivalent to a || (b && (!c))

d = true
puts "a || b == c && d != a: #{a || b == c && d != a}" # Equivalent to (a || ((b == c) && (d != a)))

puts "!a && b || c == d && d != a: #{!a && b || c == d && d != a}" # Equivalent to ((!a) && b) || ((c == d) && (d != a))

# 4. Operator associativity rules
puts "\n4. Operator associativity rules in Ruby: left-associative"
puts "a && b && c && d: #{a && b && c && d}" # Evaluates as (((a && b) && c) && d)

x = 0
y = 1
puts "x > y || y > x || y == x: #{x > y || y > x || y == x}" # Evaluates as ((x > y || y > x) || y == x)

# 5. Operand evaluation order
puts "\n5. Operand evaluation order in Ruby: left-to-right"
def first_method
  puts "First method called"
  true
end

def second_method
  puts "Second method called"
  false
end

first_method && second_method  # Output: "First method called" then "Second method called"
first_method || second_method  # Output: "First method called" then "Second method called"

# 6. Short-circuit evaluation
puts "\n6. Short-circuit evaluation:"
# Ruby uses short-circuit evaluation for && and ||.

# For &&, if the first operand is false, the second operand is not evaluated.
puts false && second_method   # "Second method called" is not printed

# For ||, if the first operand is true, the second operand is not evaluated.
puts true || second_method    # "Second method called" is not printed

result = b && (a || check())
puts "Result of b && (a || check()): #{result}" # check() is not called

def check
    puts "check() function called"
    true
end
